aws/rust-runtime/aws-config/src/imds/credentials.rs

```diff
@@ -0,0 +1,487 @@
+                ..
+            }) => {
+                tracing::info!(
+                    "received 404 from IMDS when loading profile information. \
+                Hint: This instance may not have an IAM role associated."
+                );
+                Err(CredentialsError::CredentialsNotLoaded)
+            }
+            Err(other) => Err(CredentialsError::ProviderError(other.into())),
+        }
+    }
+
+    async fn credentials(&self) -> credentials::Result {
+        if self.imds_disabled() {
+            tracing::debug!("IMDS disabled by environment variable");
+            return Err(CredentialsError::CredentialsNotLoaded);
+        }
+        let get_profile = self.get_profile_uncached();
+        let profile = self.profile.get_or_try_init(|| get_profile).await?;
+        let credentials = self
+            .client()
+            .await?
+            .get(&format!(
+                "/latest/meta-data/iam/security-credentials/{}",
+                profile
+            ))
+            .await
+            .map_err(|e| CredentialsError::ProviderError(e.into()))?;
+        match parse_imds_credentials(&credentials) {
+            Ok(ImdsCredentialsResponse::Success {
+                access_key_id,
+                secret_access_key,
+                session_token,
+                expiration,
+                ..
+            }) => Ok(Credentials::new(
+                access_key_id,
+                secret_access_key,
+                session_token.map(|tok| tok.to_string()),
+                Some(expiration),
+                "IMDSv2",
+            )),
+            Ok(ImdsCredentialsResponse::Error { code, message })
+                if code == codes::ASSUME_ROLE_UNAUTHORIZED_ACCESS =>
+            {
+                Err(CredentialsError::InvalidConfiguration(
+                    format!(
+                        "Incorrect IMDS/IAM configuration: [{}] {}. \
+                        Hint: Does this role have a trust relationship with EC2?",
+                        code, message
+                    )
+                    .into(),
+                ))
+            }
+            Ok(ImdsCredentialsResponse::Error { code, message }) => {
+                Err(CredentialsError::ProviderError(
+                    format!(
+                        "Error retrieving credentials from IMDS: {} {}",
+                        code, message
+                    )
+                    .into(),
+                ))
+            }
+            // got bad data from IMDS, should not occur during normal operation:
+            Err(invalid) => Err(CredentialsError::Unhandled(invalid.into())),
+        }
+    }
+}
+
+/// Internal data mapping for IMDS response
+#[derive(PartialEq, Eq, Debug)]
+enum ImdsCredentialsResponse<'a> {
+    Success {
+        access_key_id: Cow<'a, str>,
+        secret_access_key: Cow<'a, str>,
+        session_token: Option<Cow<'a, str>>,
+        expiration: SystemTime,
+        tpe: Cow<'a, str>,
```

Comment:
I read `tpe` as an acronym and wondered what it stood for before realizing it was working around the `type` reserved word. Maybe call it `r#type`, `variety`, `algorithm`, or `kind`?